src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -0,0 +1,224 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    /// <summary>
+    /// Provides a wrapper around the ZLib compression API.
+    /// </summary>
+    internal sealed class WebSocketDeflater : IDisposable
+    {
+        private ZLibStreamHandle? _stream;
+        private readonly int _windowBits;
+        private readonly bool _persisted;
+
+        private byte[]? _buffer;
+
+        internal WebSocketDeflater(int windowBits, bool persisted)
+        {
+            Debug.Assert(windowBits >= 9 && windowBits <= 15);
+
+            // We use negative window bits in order to produce raw deflate data
+            _windowBits = -windowBits;
+            _persisted = persisted;
+        }
+
+        public void Dispose() => _stream?.Dispose();
+
+        public void ReleaseBuffer()
+        {
+            if (_buffer is not null)
+            {
+                ArrayPool<byte>.Shared.Return(_buffer);
+                _buffer = null;
+            }
+        }
+
+        public ReadOnlySpan<byte> Deflate(ReadOnlySpan<byte> payload, bool continuation, bool endOfMessage)
+        {
+            Debug.Assert(_buffer is null, "Invalid state, ReleaseBuffer not called.");
+
+            // Do not try to rent more than 1MB initially, because it will actually allocate
+            // instead of renting. Be optimistic that what we're sending is actually going to fit.
+            const int MaxInitialBufferLength = 1024 * 1024;
+
+            // For small payloads there might actually be overhead in the compression and the resulting
+            // output might be larger than the payload. This is why we rent at least 4KB initially.
+            const int MinInitialBufferLength = 4 * 1024;
+
+            _buffer = ArrayPool<byte>.Shared.Rent(Math.Clamp(payload.Length, MinInitialBufferLength, MaxInitialBufferLength));
+            int position = 0;
+
+            while (true)
+            {
+                DeflatePrivate(payload, _buffer.AsSpan(position), continuation, endOfMessage,
+                    out int consumed, out int written, out bool needsMoreOutput);
+                position += written;
+
+                if (!needsMoreOutput)
+                {
+                    break;
+                }
+
+                payload = payload.Slice(consumed);
+
+                // Rent a 30% bigger buffer
+                byte[] newBuffer = ArrayPool<byte>.Shared.Rent((int)(_buffer.Length * 1.3));
```

Comment:
What seems dangerous to me here, is that if we've rented the biggest buffer (read 1MB) at line 54, we will grow past 1MB here and we don't even have any upper limit anymore. If the user gave us a buffer that's gigabytes in size and it's badly compressible, we would end up allocating a lot. 

I understand the reasoning to minimize the IO, but we should still have some upper bound, after which we'd want to split into several messages. It may be this 1MB limit, or possibly less, @scalablecory says that after 64K we shouldn't see effects of IO anymore on one message vs several.

Reply:
@CarnaViire we have this problem today with websockets https://github.com/dotnet/runtime/blob/3870c9f7f85df78aa12e48e0157998d8f4dc233a/src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs#L534

Reply:
@CarnaViire this is doable, but as @davidfowl indicated - this will need to be changed there as well. Can I suggest we do this with a separate PR, so we can take the time and measure what a good max buffer segment length would be? This would also require us to change the implementation of the `SendFrameAsync` where right now we only have one write to the underlying NetworkStream.

Also this change would be significantly easier to make if this gets approved at some point: https://github.com/dotnet/runtime/issues/25344.

Reply:
I'm more with @zlatanov on the idea to do it in a separate PR... This one is complex enough already. And we would need extensive perf research to make sure both non-compressed and compressed routes will not regress. For this PR, we agreed that we'd like to touch existing non-compressing code as little as possible to avoid regressions.
Also, such big messages should be rare. We should commit to still fixing that though and create an issue to track it.

P.S.: I've discussed it offline with @scalablecory and he is not against it.